### PR TITLE
Enable pandas-style rounding of cftime.datetime objects

### PR DESCRIPTION
**Pull request by spencerkclark**
- [x] Tests added
 - [x] Passes `isort -rc . && black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

This is particularly useful for removing microsecond noise that can sometimes be added from decoding times via `cftime.num2date`, though also applies more generally.  The methods used here for rounding dates in the integer domain are copied from pandas.

On a somewhat more internal note, this adds an `asi8` property to `CFTimeIndex`, which encodes the dates as integer values representing microseconds since 1970-01-01; this encoding is made exact via the `exact_cftime_datetime_difference` function.  It's possible this could be useful in other contexts.

Some examples:

```
In [1]: import xarray as xr

In [2]: times = xr.cftime_range("2000", periods=5, freq="17D")

In [3]: time = xr.DataArray(times, dims=["time"], name="time")

In [4]: time.dt.floor("11D")
Out[4]:
<xarray.DataArray 'floor' (time: 5)>
array([cftime.DatetimeGregorian(1999-12-31 00:00:00),
       cftime.DatetimeGregorian(2000-01-11 00:00:00),
       cftime.DatetimeGregorian(2000-02-02 00:00:00),
       cftime.DatetimeGregorian(2000-02-13 00:00:00),
       cftime.DatetimeGregorian(2000-03-06 00:00:00)], dtype=object)
Coordinates:
  * time     (time) object 2000-01-01 00:00:00 ... 2000-03-09 00:00:00

In [5]: time.dt.ceil("11D")
Out[5]:
<xarray.DataArray 'ceil' (time: 5)>
array([cftime.DatetimeGregorian(2000-01-11 00:00:00),
       cftime.DatetimeGregorian(2000-01-22 00:00:00),
       cftime.DatetimeGregorian(2000-02-13 00:00:00),
       cftime.DatetimeGregorian(2000-02-24 00:00:00),
       cftime.DatetimeGregorian(2000-03-17 00:00:00)], dtype=object)
Coordinates:
  * time     (time) object 2000-01-01 00:00:00 ... 2000-03-09 00:00:00

In [6]: time.dt.round("11D")
Out[6]:
<xarray.DataArray 'round' (time: 5)>
array([cftime.DatetimeGregorian(1999-12-31 00:00:00),
       cftime.DatetimeGregorian(2000-01-22 00:00:00),
       cftime.DatetimeGregorian(2000-02-02 00:00:00),
       cftime.DatetimeGregorian(2000-02-24 00:00:00),
       cftime.DatetimeGregorian(2000-03-06 00:00:00)], dtype=object)
Coordinates:
  * time     (time) object 2000-01-01 00:00:00 ... 2000-03-09 00:00:00
```

### CODE REVIEW

File: xarray/coding/cftimeindex.py [hunk 1]
```diff
@@ -528,6 +528,83 @@ def strftime(self, date_format):
         """
         return pd.Index([date.strftime(date_format) for date in self._data])
 
+    @property
+    def asi8(self):
```
Comment (dcherian) [inline xarray/coding/cftimeindex.py]: This does seem useful. Shall we list it as public API?

File: xarray/tests/test_accessor_dt.py [hunk 2]
```diff
@@ -435,3 +436,94 @@ def test_seasons(cftime_date_type):
...
+        [
+            [cftime_date_type(1, 1, 1, 0), cftime_date_type(1, 1, 1, 0)],
+            [cftime_date_type(1, 1, 1, 0), cftime_date_type(1, 1, 2, 0)],
+        ],
+        name="floor",
+    )
+
+    if use_dask:
+        chunks = {"dim_0": 1}
+        result = cftime_rounding_dataarray.chunk(chunks).dt.floor(freq)
```
Comment (dcherian) [inline xarray/tests/test_accessor_dt.py]: Can we surround this with `raise_if_dask_computes` imported from `test_dask.py` just to be sure no computes happen indirectly?

File: xarray/coding/cftimeindex.py (hunk 1 above)
Comment (spencerkclark) [reply xarray/coding/cftimeindex.py]: Yeah, that sounds like a good idea.  It looks like it is [already listed in the docs for `CFTimeIndex`](http://xarray.pydata.org/en/stable/generated/xarray.CFTimeIndex.html) (despite the fact that it currently returns `None`); is there anything else that I would need to do?  

I added a couple tests for good measure (including one that turned up a floating point issue that I fixed).

File: xarray/tests/test_accessor_dt.py (hunk 2 above)
Comment (spencerkclark) [reply xarray/tests/test_accessor_dt.py]: Good idea.  It happens that we do one minor compute to check if the array contains `cftime.datetime` objects before proceeding through the `dt` accessor; I added a comment noting this in the tests.

**Comment by dcherian** [comment]
Thanks @spencerkclark

**Pull request merged**